pkg/cluster/bootstrap.go

```diff
@@ -82,19 +85,138 @@ func (c *Cluster) shouldBootstrapLoad(ctx context.Context) (bool, error) {
 		}
 	}
 
+	if err := c.validateBootstrapCertificates(); err != nil {
+		return true, nil
+	}
+
 	// Check the stamp file to see if we have successfully bootstrapped using this token.
 	// NOTE: The fact that we use a hash of the token to generate the stamp
 	//       means that it is unsafe to use the same token for multiple clusters.
 	stamp := c.bootstrapStamp()
 	if _, err := os.Stat(stamp); err == nil {
 		logrus.Info("Cluster bootstrap already complete")
-		return false, nil
+		return true, nil
 	}
 
 	// No errors and no bootstrap stamp, need to bootstrap.
 	return true, nil
 }
 
+// isDirEmpty checks to see if the given directory
+// is empty.
+func isDirEmpty(name string) (bool, error) {
+	f, err := os.Open(name)
+	if err != nil {
+		return false, err
+	}
+	defer f.Close()
+
+	_, err = f.Readdir(1)
+	if err == io.EOF {
+		return true, nil
+	}
+	return false, err
+}
+
+func (c *Cluster) validateBootstrapCertificates() error {
+	bootstrapDirs := []string{
+		"tls",
+		"tls/etcd",
+	}
+
+	for _, dir := range bootstrapDirs {
+		if _, err := os.Stat(filepath.Join(c.config.DataDir, dir)); os.IsNotExist(err) {
+			logrus.Debugf("missing %s directory from ${data-dir}", dir)
+			return fmt.Errorf("missing %s directory from ${data-dir}", dir)
+		}
+
+		ok, err := isDirEmpty(filepath.Join(c.config.DataDir, dir))
+		if err != nil {
+			return err
+		}
+		if ok {
+			logrus.Debugf("%s directory is empty", dir)
+			return fmt.Errorf("%s directory is empty", dir)
+		}
+	}
+
+	// check existence of certificate and contents against known bootstrap data and
+	// if there are any differences, return an error to trigger a rebootstrap
+	bootstrapCertificateAndFile := map[string]string{
+		c.config.Runtime.ControlRuntimeBootstrap.ETCDServerCA:    "etcd/server-ca.crt",
```

Comment:
Do we need to gate checking these on the cluster having managed etcd enabled?

Reply:
I see that the original PR to add managed etcd support has us create the etcd certs unconditionally, even if we're not using managed etc. I'll open a separate issue for this.
Ref: https://github.com/k3s-io/k3s/commit/39571424dda6a177abd8205dcc67c5c05dbfc70a